flink-ml-iteration/flink-ml-iteration-common/src/main/java/org/apache/flink/iteration/operator/feedback/MpscQueue.java

```diff
@@ -0,0 +1,128 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.iteration.operator.feedback;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.runtime.io.disk.iomanager.IOManager;
+import org.apache.flink.runtime.memory.MemoryAllocationException;
+import org.apache.flink.runtime.memory.MemoryManager;
+import org.apache.flink.runtime.util.EmptyMutableObjectIterator;
+import org.apache.flink.statefun.flink.core.queue.Lock;
+import org.apache.flink.statefun.flink.core.queue.Locks;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.util.MutableObjectIterator;
+import org.apache.flink.util.Preconditions;
+
+import java.io.Closeable;
+import java.io.IOException;
+
+/**
+ * Multi producers single consumer fifo queue.
+ *
+ * @param <T> The element type.
+ */
+@Internal
+public final class MpscQueue<T> implements Closeable {
+    private final Lock lock = Locks.spinLock();
+
+    private SpillableFeedbackQueue<T> activeQueue;
+    private SpillableFeedbackQueue<T> standByQueue;
+
+    private volatile boolean isClosed;
+
+    public MpscQueue(
+            IOManager ioManager,
+            MemoryManager memoryManager,
+            TypeSerializer<T> serializer,
+            long inMemoryBufferSize,
+            long pageSize)
+            throws MemoryAllocationException {
+        this.activeQueue =
+                new SpillableFeedbackQueue<>(
+                        ioManager, memoryManager, serializer, inMemoryBufferSize / 2, pageSize);
+        this.standByQueue =
+                new SpillableFeedbackQueue<>(
+                        ioManager, memoryManager, serializer, inMemoryBufferSize / 2, pageSize);
+    }
+
+    /**
+     * Adds an element to this (unbound) queue.
+     *
+     * @param element the element to add.
+     * @return the number of elements in the queue after the addition.
+     */
+    public long add(T element) {
+        Preconditions.checkState(element instanceof StreamRecord);
+
+        lock.lockUninterruptibly();
+        try {
+            SpillableFeedbackQueue<T> active = this.activeQueue;
+
+            active.add(element);
+            return active.size();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Atomically drains the queue.
+     *
+     * @return a batch of elements that obtained atomically from that queue.
+     */
+    public MutableObjectIterator<T> drainAll() {
+        lock.lockUninterruptibly();
+        try {
+            final SpillableFeedbackQueue<T> ready = this.activeQueue;
+            if (ready.size() == 0) {
+                return EmptyMutableObjectIterator.get();
+            }
+            this.activeQueue = this.standByQueue;
+            this.standByQueue = ready;
+            return ready.iterate();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public void resetStandBy() throws Exception {
```

Comment:
I agree. Now that I have changed these methods to package private, maybe it is not necessary to rename the `resetStandBy` because it is not visible to users anymore.